drivers/virtio/virtio-mmio.c

```diff
@@ -0,0 +1,226 @@
+      return ret;
+    }
+
+  /* Reset */
+
+  putreg32(0, &regs->status);
+  mb();
+
+  /* Set ack */
+
+  val = getreg32(&regs->status) | VIRTIO_STATUS_ACKNOWLEDGE;
+  putreg32(val, &regs->status);
+  mb();
+
+  /* Set driver */
+
+  val = getreg32(&regs->status) | VIRTIO_STATUS_DRIVER;
+  putreg32(val, &regs->status);
+  mb();
+
+  /* Check the device_id */
+
+  val = getreg32(&regs->device_id);
+
+  switch (val)
+    {
+#ifdef CONFIG_DRIVERS_VIRTIO_NET
+    case VIRTIO_DEV_NET:
+      ret = virtnet_init(regs, irq);
+      break;
+#endif
+    default:
+      _warn("unsupported device_id 0x%" PRIx32 "\n", val);
+    }
+
+  return ret;
+}
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: virtq_alloc_desc
+ ****************************************************************************/
+
+uint32_t virtq_alloc_desc(FAR struct virtqueue *virtq,
+                          uint32_t idx, FAR void *addr)
+{
+  uint32_t id = idx % virtq->len;
+
+  _info("virtq=%p, idx=%" PRId32 "\n", virtq, idx);
+
+  virtq->desc[id].addr = (uintptr_t)addr;
+  virtq->desc_virt[id] = addr;
+  return id;
+}
+
+/****************************************************************************
+ * Name: virtq_add_to_device
+ ****************************************************************************/
+
+void virtq_add_to_device(FAR struct virtio_mmio_regs *regs,
+                         FAR struct virtqueue *virtq,
+                         uint32_t queue_sel)
+{
+  _info("==== queue_sel=%" PRId32 ", virtq->len=%" PRId32 "\n",
+        queue_sel, virtq->len);
+
+  putreg32(queue_sel, &regs->queue_sel);
+  mb();
+
+  putreg32(virtq->len, &regs->queue_num);
+
+  putreg32((uint32_t)(ptr_to_uint64(virtq->desc)),
+           &regs->queue_desc_low);
+  putreg32((uint32_t)(ptr_to_uint64(virtq->desc) >> 32),
+           &regs->queue_desc_high);
+
+  putreg32((uint32_t)(ptr_to_uint64(virtq->avail)),
+           &regs->queue_avail_low);
+  putreg32((uint32_t)(ptr_to_uint64(virtq->avail) >> 32),
+           &regs->queue_avail_high);
+
+  putreg32((uint32_t)(ptr_to_uint64(virtq->used)),
+           &regs->queue_used_low);
+  putreg32((uint32_t)(ptr_to_uint64(virtq->used) >> 32),
+           &regs->queue_used_high);
+
+  mb();
+  putreg32(1, &regs->queue_ready);
+}
+
+/****************************************************************************
+ * Name: virtio_init
+ ****************************************************************************/
+
+void virtio_init(void)
+{
+  uintptr_t virtio = (uintptr_t)CONFIG_DRIVERS_VIRTIO_BASE;
+  uint32_t  irq = CONFIG_DRIVERS_VIRTIO_IRQ;
+  uint32_t  size = CONFIG_DRIVERS_VIRTIO_REGSIZE;
+  uint32_t  i;
+
+  for (i = 0; i < CONFIG_DRIVERS_VIRTIO_NUM; i++)
+    {
+      virtio_dev_init(virtio + size * i, irq + i);
```

Comment:
Can all devices share this rule for address and irq?

Reply:
Yes